solana/restaking-v2/programs/restaking-v2/src/lib.rs

```diff
@@ -312,6 +372,123 @@ pub mod restaking_v2 {
 
         Ok(())
     }
+
+    pub fn update_token_price(ctx: Context<UpdateTokenPrice>) -> Result<()> {
+        let common_state = &mut ctx.accounts.common_state;
+
+        let token_price_feed = &ctx.accounts.token_price_feed;
+        let sol_price_feed = &ctx.accounts.sol_price_feed;
+
+        let token_mint = ctx.accounts.token_mint.key();
+
+        let validators = common_state.validators.clone();
+
+        let staked_token = common_state
+            .whitelisted_tokens
+            .iter_mut()
+            .find(|whitelisted_token| whitelisted_token.address == token_mint);
+
+        if let Some(staked_token) = staked_token {
+            if let Some(token_feed_id) = staked_token.oracle_address.as_ref() {
+                let (token_price, sol_price) = if cfg!(feature = "mocks") {
+                    let feed_id: [u8; 32] =
+                        get_feed_id_from_hex(token_feed_id)?;
+                    let sol_price = sol_price_feed.get_price_unchecked(
+                        &get_feed_id_from_hex(SOL_PRICE_FEED_ID)?,
+                    )?;
+                    let token_price =
+                        token_price_feed.get_price_unchecked(&feed_id)?;
+                    (token_price, sol_price)
+                } else {
+                    let maximum_age_in_sec: u64 = 30;
+                    let feed_id: [u8; 32] =
+                        get_feed_id_from_hex(token_feed_id)?;
+                    let sol_price = sol_price_feed.get_price_no_older_than(
+                        &Clock::get()?,
+                        maximum_age_in_sec,
+                        &get_feed_id_from_hex(SOL_PRICE_FEED_ID)?,
+                    )?;
+                    let token_price = token_price_feed
+                        .get_price_no_older_than(
+                            &Clock::get()?,
+                            maximum_age_in_sec,
+                            &feed_id,
+                        )?;
+                    (token_price, sol_price)
+                };
+
+                let token_decimals = ctx.accounts.token_mint.decimals;
+
+                let amount_in_sol_decimals = 10u64.pow(SOL_DECIMALS as u32) /
+                    10u64.pow(token_decimals as u32);
```

Comment:
This may be zero if token_decimals is greater than sol_decimals.

Reply:
storing the value by multiplying with exponent so that we dont miss the fractions.